docs/deployment/configuration.md
---
title: Configuration
layout: default
parent: Deployment
nav_order: 2
---

# Configuration

The PCS's main settings file[^1] and Flower settings file[^2] can be modified directly or with values specified through the use of environment variables. Environment variable values may be passed directly or as the contents of a file. In the latter case, the full path of the file should be passed as a variable's value. Environment variable values may be stored and passed as [Docker Secrets](https://docs.docker.com/engine/swarm/secrets/).


### Default Values

Many of the default values used in the PCS's configuration files are designed for a Docker deployment similar to the examples in the [Docker Deployment]({{ site.url }}/deployment/docker/) section of this documentation. 

{: .important-title }
> Notation
>
> **(D)** debug mode `DEBUG=True`  
> **(P)** production mode `DEBUG=False` 

<hr>

## Django

The following variables, located in the main settings file[^1], map to various Django settings. A link to the relevant Django documentation is provided for each variable.

### General

Variable | Default | Description
---|---|---
SECRET_KEY | a long string[^3] | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#std-setting-SECRET_KEY)
DEBUG | False | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#std-setting-DEBUG) 
ALLOWED_HOSTS | [] | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#allowed-hosts)
TIME_ZONE | America/New_York | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#std-setting-TIME_ZONE)


### Cache

Variable | Default | Description
---|---|---
CACHE_LOCATION | redis://redis:6379/0 | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#location)
CACHE_TIMEOUT | **(D)** 0<br>**(P)** 300  | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#timeout)


### Database 

Variable | Default | Description
---|---|---
SQL_HOST | mariadb | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#host)
SQL_PORT | 3306 | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#port)
SQL_DATABASE | contestsuite | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#name)
SQL_USER | contestadmin | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#user)
SQL_PASSWORD | seminoles1! | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#password)
SQL_TIMEZONE | America/New_York | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#time-zone) 


### Email

Variable | Default | Description
---|---|---
EMAIL_BACKEND | **(D)** `django.core.mail.backends.console.EmailBackend`<br>**(P)** `django.core.mail.backends.smtp.EmailBackend` | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#email-backend)
EMAIL_HOST | None | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#email-host)
EMAIL_PORT | 587 | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#email-port)
EMAIL_HOST_USER | None | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#email-host-user)
EMAIL_HOST_PASSWORD | None | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#email-host-password)
EMAIL_USE_SSL | False | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#email-use-ssl)
EMAIL_USE_TLS | False | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#email-use-tls)
DEFAULT_FROM_EMAIL | acm@cs.fsu.edu | [Docs](https://docs.djangoproject.com/en/4.2/ref/settings/#default-from-email)

## Celery

The following variables, located in the main settings file[^1], map to various Celery settings. A link to the relevant Celery documentation is provided for each variable.

Variable | Default | Description
---|---|---
CELERY_BROKER | amqp://rabbitmq:5672 | [Docs](https://docs.celeryq.dev/en/stable/userguide/configuration.html#std-setting-broker_url)
CELERY_BACKEND | redis://redis:6379/1 | [Docs](https://docs.celeryq.dev/en/stable/userguide/configuration.html#std-setting-result_backend)
CELERY_TIMEZONE | America/New_York | [Docs](https://docs.celeryq.dev/en/stable/userguide/configuration.html#std-setting-timezone)

## Celery Beat

The following variables, located in the main settings file[^1], map to various Celery Beat settings. A link to the relevant Celery Beat documentation is provided for each variable.

Variable | Default | Description
---|---|---
CELERY_BEAT_SCHEDULE | `{}` | [Docs](https://docs.celeryq.dev/en/stable/userguide/configuration.html#std-setting-beat_schedule)

## Flower

The following variables, located in the Flower settings file[^2], map to various Flower settings. A link to the relevant Flower documentation is provided for each variable.

Variable | Default | Description
---|---|---
FLOWER_BROKER_API | http://rabbitmq:15672/api/vhost | [Docs](https://flower.readthedocs.io/en/latest/config.html#broker-api)
FLOWER_USER **(P)** | contestadmin | [Docs](https://flower.readthedocs.io/en/latest/config.html#basic-auth)
FLOWER_PASSWORD **(P)** | seminoles1! | [Docs](https://flower.readthedocs.io/en/latest/config.html#basic-auth)
FLOWER_COOKIE_KEY **(P)** | None | [Docs](https://flower.readthedocs.io/en/latest/config.html#cookie-secret)
FLOWER_URL_PREFIX **(P)** | flower | [Docs](https://flower.readthedocs.io/en/latest/config.html#url-prefix)

## Discord

The following variables, located in the main settings file[^1], map to various PCS specific settings required for integration with Discord.

Variable | Default | Description
---|---|---
ANNOUNCEMENT_WEBHOOK_URL | None | URL of Discord server's announcement channel webhook |

## Misc

The following variables, located in the main settings file[^1], map to miscellaneous PCS specific settings. A link to relevant documentation is provided for each variable.

Variable | Default | Description
---|---|---
DOMJUDGE_URL | https://domjudge.cs.fsu.edu/public | Full URL of the DOMjudge server. Used for PCS homepage contest server status card.
PCS_DOCS_URL | https://fsu-acm.github.io/Programming-Contest-Suite | Base URL of the project's documentation website. Used to link registration guide and other manuals.
HASHID_FIELD_SALT | a long string[^4] | [Docs](https://github.com/nshafer/django-hashid-field#hashid_field_salt) The `django-hashid-field` library is used to hash sensitive PCS database model fields.
GTAG | None | [Google Analytics site tag](https://support.google.com/analytics/answer/12002338?hl=en)

---
[^1]: `Programming-Contest-Suite/src/contestsuite/settings.py`
[^2]: `Programming-Contest-Suite/src/contestsuite/flowerconfig.py`
[^3]: `86@j2=z!=&1r_hoqboog1#*mb$jx=9mf0uw#hrs@lw&7m34sqz`
[^4]: `0s97rx*t4%68jell&lw3^)97o*kr*+*2o^(76q)ix+ilc!4ax#`


src/contestsuite/settings.py
"""
Django settings for contestsuite project.

Generated by 'django-admin startproject' using Django 2.2.5.

For more information on this file, see
https://docs.djangoproject.com/en/4.2/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/4.2/ref/settings/
"""

import os

def get_secret(key, default=None):
    """
    Simple function to read an environment variable and return a default if the variable
    does not exist. Supports values passed via Docker Secrets.
    """

    
    value = os.getenv(key, default)
    if value and os.path.isfile(value):
        with open(value) as f:
            return f.read().strip()
    return value

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/4.2/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
# Raises django's ImproperlyConfigured exception if SECRET_KEY not in os.environ

SECRET_KEY = get_secret('SECRET_KEY', '86@j2=z!=&1r_hoqboog1#*mb$jx=9mf0uw#hrs@lw&7m34sqz')

# SECURITY WARNING: don't run with debug turned on in production!

if get_secret('DEBUG'):
    DEBUG = get_secret('DEBUG') == 'True'
else:
    DEBUG = False


# Allowed Hosts and Origins

ALLOWED_HOSTS = []
if DEBUG:
    ALLOWED_HOSTS = ['localhost', '127.0.0.1', '0.0.0.0', '[::1]']

if os.getenv('ALLOWED_HOSTS'):
    ALLOWED_HOSTS = ALLOWED_HOSTS + get_secret('ALLOWED_HOSTS').split(',')
    CSRF_TRUSTED_ORIGINS = [
        f"https://{hostname}" if 'https://' not in hostname else hostname for hostname in ALLOWED_HOSTS]


# Application definition

INSTALLED_APPS = [
    # Django built-ins
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    # User defined
    'announcements.apps.AnnouncementsConfig',
    'checkin.apps.CheckinConfig',
    'contestadmin.apps.ContestAdminConfig',
    'core.apps.CoreConfig',
    'lfg.apps.LfgConfig',
    'manager.apps.ManagerConfig',
    'register.apps.RegisterConfig',
    # 3rd party packages
    'django_celery_beat',
    'import_export',
]


MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]


ROOT_URLCONF = 'contestsuite.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, 'templates')],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                #'django.template.context_processors.debug', # replaced by contestsuite.context_processors.app_settings
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
                'contestsuite.context_processors.app_settings',
            ],
            'libraries': {
                'custom_filters': 'templatetags.custom_filters',
            },
        },
    },
]

WSGI_APPLICATION = 'contestsuite.wsgi.application'


# Database
# https://docs.djangoproject.com/en/4.2/ref/settings/#databases

# Parse database connection url strings like psql://user:pass@127.0.0.1:8458/db
# read os.environ['DATABASE_URL'] and raises ImproperlyConfigured exception if not found

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.mysql',
        'HOST': get_secret('SQL_HOST', 'mariadb'),
        'PORT': get_secret('SQL_PORT', '3306'),
        'NAME': get_secret('SQL_DATABASE', 'contestsuite'),
        'USER': get_secret('SQL_USER', 'contestadmin'),
        'PASSWORD': get_secret('SQL_PASSWORD', 'seminoles1!'),
        'OPTIONS': {'charset': 'utf8mb4'},
        'TIME_ZONE': get_secret('SQL_TIMEZONE', 'America/New_York'),
        'AUTOCOMMIT': True,
        'CONN_MAX_AGE': 0,
    }
}

DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'


# Cache
# https://docs.djangoproject.com/en/4.2/ref/settings/#caches

CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.redis.RedisCache',
        'LOCATION': get_secret('CACHE_LOCATION', 'redis://redis:6379/0'),
    }
}

if DEBUG:
    CACHE_TIMEOUT = int(get_secret('CACHE_TIMEOUT', 0))
else:
    CACHE_TIMEOUT = int(get_secret('CACHE_TIMEOUT', 300))


# Password validation
# https://docs.djangoproject.com/en/4.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/4.2/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = get_secret('TIME_ZONE', 'America/New_York')

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/4.2/howto/static-files/

STATIC_ROOT = os.path.join(BASE_DIR, 'static')
STATIC_URL = '/static/'
STATICFILES_DIRS = [
    os.path.join(BASE_DIR, 'assets'),
]


# Non-static files
# https://docs.djangoproject.com/en/4.2/topics/files/

MEDIA_ROOT = os.path.join(BASE_DIR, 'media')
MEDIA_URL = '/media/'


# Redirect to home URL after login (Default redirects to /accounts/profile/)

LOGIN_REDIRECT_URL = '/manage/'


# Sessions
# https://docs.djangoproject.com/en/4.2/topics/http/sessions/

SESSION_ENGINE = 'django.contrib.sessions.backends.cache'

if not DEBUG:
    SESSION_EXPIRE_AT_BROWSER_CLOSE = True


# Messages
# https://docs.djangoproject.com/en/4.2/ref/contrib/messages/

MESSAGE_STORAGE = 'django.contrib.messages.storage.session.SessionStorage'


# Email
# https://docs.djangoproject.com/en/4.2/topics/email/

if DEBUG:
    EMAIL_BACKEND = get_secret('EMAIL_BACKEND', 'django.core.mail.backends.console.EmailBackend')
else:
    EMAIL_BACKEND = get_secret('EMAIL_BACKEND', 'django.core.mail.backends.smtp.EmailBackend')
      
EMAIL_HOST = get_secret('EMAIL_HOST', None)
EMAIL_PORT = int(get_secret('EMAIL_PORT', 587))
EMAIL_HOST_USER = get_secret('EMAIL_USER', None)
EMAIL_HOST_PASSWORD = get_secret('EMAIL_PASSWORD', None)

if get_secret('EMAIL_USE_SSL'):
    EMAIL_USE_SSL = get_secret('EMAIL_USE_SSL') == 'True'
else:
    EMAIL_USE_SSL = False

if get_secret('EMAIL_USE_TLS'):
    EMAIL_USE_TLS = get_secret('EMAIL_USE_TLS') == 'True'
else:
    EMAIL_USE_TLS = False

DEFAULT_FROM_EMAIL = get_secret(
    'DEFAULT_FROM_EMAIL', 'ACM at FSU Programming Contest<acm@cs.fsu.edu>')


# Celery
# https://docs.celeryproject.org/en/stable/getting-started/first-steps-with-celery.html#configuration
 
CELERY_BROKER_URL = get_secret('CELERY_BROKER', 'amqp://rabbitmq:5672')
CELERY_RESULT_BACKEND = get_secret('CELERY_BACKEND', 'redis://redis:6379/1')
CELERY_TASK_SERIALIZER = 'json'
CELERY_RESULT_SERIALIZER = 'json'
CELERY_ACCEPT_CONTENT = ['json']
CELERY_TIMEZONE = get_secret('CELERY_TIMEZONE', 'America/New_York')
CELERY_ENABLE_UTC = True


# Celery Beat
# https://celery-safwan.readthedocs.io/en/latest/reference/celery.beat.html

CELERY_BEAT_SCHEDULE = {}


# Discord
# https://discordpy.readthedocs.io/en/stable/

ANNOUNCEMENT_WEBHOOK_URL = get_secret('ANNOUNCEMENT_WEBHOOK_URL', None)


# Hashid Fields
# https://pypi.org/project/django-hashid-field/

HASHID_FIELD_SALT = get_secret(
    'HASHID_FIELD_SALT', '0s97rx*t4%68jell&lw3^)97o*kr*+*2o^(76q)ix+ilc!4ax#')


# Google Analytics
# https://support.google.com/analytics/answer/12002338?hl=en

GTAG = get_secret('GTAG')


# DOMjudge status button on homepage

DOMJUDGE_URL = get_secret('DOMJUDGE_URL', 'https://domjudge.cs.fsu.edu/public')


# Documentation website base URL for navbar and homepage documentation links

PCS_DOCS_URL = get_secret('PCS_DOCS_URL', 'https://fsu-acm.github.io/Programming-Contest-Suite')
